### PR TITLE
[IMP] web_editor: open internal link in edit mode

### PR DESCRIPTION
**Pull request by divy-odoo**
Before this commit, when the user was clicking on a link to another page
of the website, it was opening in another tab but in frontend mode.
However, in most cases the user actually wanted it to open in backend
mode.

Here is an example of users who started an eCommerce trial. They are
editing their homepage and want to create a product. They open the
"/shop" page using popover but because they are in the frontend, they
don't see the "+NEW" button. This behaviour seems a little odd.

Therefore, this commit aims to open internal links in the edit(backend)
mode instead of the frontend mode.

task-3177126

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js:61]: This doesn't belong in `web_editor`, this `/@` controller is a website one, meaning that this is just going to redirect you to 404 urls when website is not installed.

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js:196]: why do we want to use the `/@/` for the preview?

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js:60]: I doubt we want to change that and copy the url with `/@/` inside it..
The goal is to reuse the link in another link, and we don't want user to copy paste and use link with `/@/` inside it

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:22]: Many linter error please be sure to have a proper js linter

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:23]: const, not let

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:25]: What if the URL already has a `/@/` inside it?
image

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:27]: So here this is a text url, but the other case it is a `new URL()`, does it even work when you later do 
```
const clipboard = new _ClipboardJS(
            this.$copyLink[0],
            {text: () => this.url}
```
?

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:24]: `url.protocol.startsWith('http')` ?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Seems overkill and wrong approach.
The spec is just to make clickable links of the preview (which open in new tab) to redirect to the backend version (`/@`).

But here it's changing way more than just that: the Clipboard, the preview itself..

IMHO, just add a click handler on those links (instead of natural href+_blank) and do the opening yourself with ensuring the `/@` is here, using `pathname.startsWith('/@/')` and url concat.

Things to consider:
- URL might already have `/@` inside it -> here you are adding it twice
- Relative/Absolute

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:25]: image

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:33]: image

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the changes but this won't be reviewed until there is a configured linter, as previously requested here https://github.com/odoo/odoo/pull/121264#discussion_r1242016982

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:26]:
```suggestion
     * Opens website page links in backend mode by forcing the '/@/' controller.
```
- (Important) Max 80 char
- (Ideally) Start with a verb (with `s` at the end)
- (Nitpicking) We know we are in edit mode, this file is for link preview which is an edit mode feature

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js:191]: `click` is the default `run` so in this case you would simply remove it.
But I'd like to have @detrouxdev opinion about this, because he wrote that step on purpose (see https://github.com/odoo/odoo/commit/f9e51a8ab1cc469bd3dd4dba21856597d5e0f902) but I don't understand what it does.

It's supposed to fix a runbot race condition so I don't want to revive it

I saw one of your message on that PR https://github.com/odoo/odoo/pull/114470#issuecomment-1463907491:
> About fixing that however, I guess there's multiple ways of handling it:
>  [...]
> Change the link popover clicks **to not rely on the default behaviour**

It's what is done here.

**Review by rdeodoo — APPROVED** [review]
Except for the test for which I am unsure, this is LGTM.
I'll wait for ARD feedback before merging it.

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (detrouxdev) [reply addons/website/static/tests/tours/edit_link_popover.js:191]: The step is in reference to [this comment](https://github.com/odoo/odoo/pull/114470#pullrequestreview-1333182991), more specifically:

> Could be nice to have a test to enforce this... although not sure how we would detect a link being opened in a new tab

The goal was to find a way to ensure a tab was opened after clicking the link. We settled on assuming that if the click event was not "preventDefault"ed, it would open on a tab.

However here, with this change, this line would not test anything. As it's impossible to check the amount of tabs that are open in the headless chrome browser. (It might be possible Python side but I don't think the test framework is made for that, might be worth checking?).

We might be able to check if the window.open method was properly called though. If instead of directly calling window.open you import browser from "@web/core/browser/browser" and in the test you patch that object to replace the open method with one that checks if the URL is the correct one. You can check `rpc_services_test.js` for an example, though it's done in QUnit, I'm pretty sure we could find a way to do it in the tour.

It might be overkilled but I think it's worth testing these kind of features as we could easily overlook it.

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js:191]: Thanks for the explanation :+1: @divy-odoo could you do that please?

Note:

>  We settled on assuming that if the click event was not "preventDefault"ed, it would open on a tab.

Or it could also mean that it was navigating in the current tab, no?

Comment (detrouxdev) [reply addons/website/static/tests/tours/edit_link_popover.js:191]: Yup, I should have added a step to ensure that we stayed on the same page. That's an oversight 😬

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (divy-odoo) [inline addons/website/static/tests/tours/edit_link_popover.js:204]: Hello @detrouxdev @rdeodoo,
Here is the implementation of the way you suggested.

Before this, the test was focused on testing that a link should open in a new window. 
But IMO now, the test focus should be on testing that link should open in backend mode, no?

There are 3 cases we have,
 
1. open a link in a new window in a tour.
I have doubts about this, do we need to open a link during the tour?
If yes, we can do something like this by calling `window.open(URL)`.
new_window

2. Open a link in the same window in a tour.
Is it really will be good to change the actual work of function as the tour will be a guide for the user? like, when the user clicks on the link it will redirect to the new window & while in the tour, it won't. Is this desired behaviour? 
Also, it will create an issue when there are more tour steps we will have after this. ( as the page will be refreshed after this tour step )

3. Just test that a link URL is correct, don't open the link. ( which I have implemented currently )
In this, ensure that the link should open in edit mode without opening it. 

What are your thoughts on it?

Comment (detrouxdev) [reply addons/website/static/tests/tours/edit_link_popover.js:204]: This is a test tour so there's no need to worry about the user! It's just a way for us to programatically test the UI.

I think each option has a valid positive to it.

1. This would not modify the behaviour and would ensure we're as close to reality as possible
2. We could technically redirect but I think this is not a good idea and could lead to race conditions.
3. IMO this is the best option, you check the URL and it's good enough for me.

I don't think you should patch and unpatch in the same step, I'm not 100% sure, but there's a risk that you unpatch while the event handler hasn't even been processed yet. Instead you could modify the DOM in the patched open method and check if the DOM is properly modified in the next step. 😄

Comment (divy-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js:204]: Ok!! then good to go with option 3.

But do we really need to force modify the DOM, as the link click method will not have any DOM modification inside?
Agree with you @detrouxdev  there can be a risk, but that can be handled by making the method `async` too, I guess. no?

Something like, 
async_function

Comment (divy-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js:204]: > somehow await for the open method to be called

:thinking:  Ok @detrouxdev. 

Hope this is what we are expecting, 
- it will resolve the promise as soon as the open method is executed and allowed to unpatch.

```javascript
run: async function (actions) {
    patch(browser, 'window_open_action', {
        open: (url) => {
            if (window.location.hostname === url.hostname && !url.pathname.startsWith('/@/')) {
                throw new Error("The link popover should redirect to the backend link.");
            }
            this.resPromise("The link popover will open link in backend.") 
        }
    }, { pure: true }) // We do not want to open a link in a tour
    await new Promise((resolve) => {
        this.resPromise = resolve
        actions.click();
    });
    unpatch(browser,'window_open_action')
}

// ( due to having an issue with variable scoping, I have to take this.resPormise variable ) 
```
What do think about this?

Comment (detrouxdev) [reply addons/website/static/tests/tours/edit_link_popover.js:204]: If it works, it's okay with me.
Though I would not put the resolve inside this, just declare an empty variable with
`let resPromise`; at the top of the run method. You should have access to it in both your functions.

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: @divy-odoo Are you sure that the `run()` is really awaited? I always thought it wasn't.
IMO, the test will finish before executing (fully?) this run javascript code.
I would expect the promise success to set some attribute in the DOM and a next step to check that attribute.
So, if the tour moves to the next step while this JS is not really finished, the step will not succeed until the JS did what is expected and the test is correctly tested.

From what I'd imagine, the test is:
- Checking the trigger and extra trigger of your step, which succeed
- Starting to run the custom js `Run` of that step
- Directly checking the next step triggers, which if found will succeed regardless of the previous run js code status
  But in this case there is no step after that anyway, so I expect the tour to be considered successful and finish at the same time as the run is executed.

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: I actually gave it a test, and it's how I expected:
```
Tour Manager is ready.  running_tour=test_unsplash_beacon
TIP.RUN IS A FUNCTION, GOIGN TO CALL THE FUNCTION
body found, starting 4s sleep
Tour test_unsplash_beacon: step 'Verify whether beacon was sent. (trigger: body)' succeeded
TIP.RUN IS A FUNCTION, GOIGN TO CALL THE FUNCTION
body found (2), starting 4s sleep
Tour test_unsplash_beacon: step 'Verify whether beacon was sent. (trigger: body)' succeeded
Tour test_unsplash_beacon succeeded
test successful
end sleep
end sleep
```

This is with the following code:
```js
tour.register("test_unsplash_beacon", {
    test: true,
    url: "/?test_unsplash_beacon",
}, [{
    content: "Verify whether beacon was sent.",
    trigger: 'body',
    run: async () => {
        console.log('body found, starting 4s sleep');
        await sleep(4000);
        console.log('end sleep');
    },
},
{
    content: "Verify whether beacon was sent.",
    trigger: 'body',
    run: async () => {
        console.log('body found (2), starting 4s sleep');
        await sleep(4000);
        console.log('end sleep');
    },
}]);
});
```

Comment (detrouxdev) [reply addons/website/static/tests/tours/edit_link_popover.js]: Ah I spend too much time in master I forgot when they converted the tour util into the new "macros" ( [done in 16.3](https://github.com/odoo/odoo/commit/3a798039d6f200f8e28448ddb6a2d3c46654a203) )
In master with the same code:
```
Tour test_unsplash_beacon on step: 'Verify whether beacon was sent. (trigger: body)'
unsplash_beacon.js:31 body found, starting 4s sleep
unsplash_beacon.js:33 end sleep
tour_compilers.js:295 Tour test_unsplash_beacon on step: 'Verify whether beacon was sent. (trigger: body)'
unsplash_beacon.js:40 body found (2), starting 4s sleep
unsplash_beacon.js:42 end sleep
tour_service.js:215 test successful
```

Doesn't matter much here but just thought it was interesting to note.

Pretty sure you will have to modify the DOM in the patched open function and add a step then @divy-odoo, at least up to <16.3

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: @detrouxdev Nice, thank you for the precision.
@divy-odoo Just add a class to the body, no need of a new element :+1:

Comment (divy-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: Hello @rdeodoo, The changes as been applied as per the discussion.

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: const not let

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: run()

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: It's not a check. Check is for empty run when you just want to check for the existence of the trigger without clicking on it, here the run is not empty

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: It's useless to remove it, what is done in the test is removed/deleted after the test, so this is not needed (and not even awaited anyway)

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: What does this comment have to do with the `pure: true` option?

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: const

Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: When you have a variable holding an html element, suffix it with `El`

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: if the patch fail, the class won't be added, and the next step will fail, this is not needed IMO

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: The async does nothing as we already discussed here https://github.com/odoo/odoo/pull/121264#discussion_r1254226862
The `run()` is not async capable.

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: Do we really need to unpatch?

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (rdeodoo) [inline addons/website/static/tests/tours/edit_link_popover.js]: linter

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js:32]: Missing separator
```js
    //--------------------------------------------------------------------------
    // Handlers
    //--------------------------------------------------------------------------
```

File: addons/website/static/src/js/widgets/link_popover_widget.js
Comment (rdeodoo) [inline addons/website/static/src/js/widgets/link_popover_widget.js]: New line between imports and business code, here you removed it by typing something inside it

**Review by rdeodoo — APPROVED** [review]
@robodoo r+ 

I have rewritten the diff of the test, most was useless it seems.
I also corrected all remarks myself so we can ship it without waiting another round of review/change.
Here is the [diff](https://github.com/odoo/odoo/compare/970d2baa2f2c8c583750f069c82640e26b383c51..c18ded99239784eb6848c541f97651ad84aae076).

File: addons/website/static/tests/tours/edit_link_popover.js
Comment (divy-odoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: IMO Yes!! @rdeodoo :thinking: 

> It's useless to remove it, what is done in the test is removed/deleted after the test, so this is not needed (and not even awaited anyway)

IDK if it's expected behaviour or not but,
The changes made during the test are getting destroyed once we refresh the page, not after the successful test.
( refer to the attachment )

Do try, 

1. Run the test successfully
2. Click on the popover link ( Do not refresh the page ) 
-> The patch will execute.
3. Refresh the page & click on the popover link
-> Works fine.

I tried to cover the case if the tour runs manually then operations performed without refreshing are not getting blocked. 

If it's okay with this behaviour then there is no need to unpatch.
`async` was to handle the `unpatch()`, If we unpatch then it's needed else no means of that too. 

test_successful

Comment (rdeodoo) [reply addons/website/static/tests/tours/edit_link_popover.js]: We don't care about running the test locally, it's just a nice to have feature for debugging purpose but this should never be considered when writing a test.
The only thing that needs to be considered is the test being launched through the testing suite on chrome headless in a transaction where everyone is rollbacked after the test is done.

**Pull request closed without merging**
